pkg/checks/traceroute/traceroute.go

```diff
@@ -0,0 +1,227 @@
+	c.Mu.Lock()
+	defer c.Mu.Unlock()
+	return &c.config
+}
+
+func (c *Traceroute) check(ctx context.Context) map[string]Result {
+	res := make(map[string]Result)
+	log := logger.FromContext(ctx)
+
+	type internalResult struct {
+		addr string
+		res  Result
+	}
+
+	var wg sync.WaitGroup
+	cResult := make(chan internalResult, len(c.config.Targets))
+
+	for _, t := range c.config.Targets {
+		wg.Add(1)
+		go func(t Target) {
+			defer wg.Done()
+			log.Debug("Running traceroute", "target", t.Addr)
+			start := time.Now()
+			tr, trerr := c.traceroute(t.Addr, int(t.Port), int(c.config.Timeout/time.Millisecond), c.config.Retries, c.config.MaxHops)
+			duration := time.Since(start)
+			if trerr != nil {
+				log.Error("Error running traceroute", "err", trerr, "target", t.Addr)
+			}
+
+			log.Debug("Ran traceroute", "result", tr, "duration", duration)
+
+			result := Result{
+				NumHops: len(tr.Hops),
+				Hops:    []Hop{},
+			}
+
+			for _, h := range tr.Hops {
+				result.Hops = append(result.Hops, Hop{
+					Addr:    h.Host,
+					Latency: h.ElapsedTime,
+					Success: h.Success,
+				})
+			}
+			cResult <- internalResult{addr: t.Addr, res: result}
+		}(t)
+	}
+
+	log.Debug("Waiting for traceroute checks to finish")
+
+	go func() {
+		wg.Wait()
+		close(cResult)
+	}()
+
+	log.Debug("All traceroute checks finished")
+
+	for r := range cResult {
+		res[r.addr] = r.res
+	}
+
+	log.Debug("Getting errors from traceroute checks")
+
+	log.Debug("Finished traceroute checks", "result", res)
+
+	return res
+}
+
+// Shutdown is called once when the check is unregistered or sparrow shuts down
+func (c *Traceroute) Shutdown(_ context.Context) error {
+	c.DoneChan <- struct{}{}
+	close(c.DoneChan)
+	return nil
+}
+
+// SetConfig is called once when the check is registered
+// This is also called while the check is running, if the remote config is updated
+// This should return an error if the config is invalid
+func (c *Traceroute) SetConfig(cfg checks.Runtime) error {
+	if cfg, ok := cfg.(*Config); ok {
+		c.Mu.Lock()
+		defer c.Mu.Unlock()
+		c.config = *cfg
+		return nil
+	}
+
+	return checks.ErrConfigMismatch{
+		Expected: CheckName,
+		Current:  cfg.For(),
+	}
+}
+
+// Schema returns an openapi3.SchemaRef of the result type returned by the check
+func (c *Traceroute) Schema() (*openapi3.SchemaRef, error) {
+	return checks.OpenapiFromPerfData[Result](Result{})
+}
+
+// GetMetricCollectors allows the check to provide prometheus metric collectors
+func (c *Traceroute) GetMetricCollectors() []prometheus.Collector {
+	return []prometheus.Collector{}
+}
+
+func (c *Traceroute) Name() string {
+	return CheckName
+}
+
+func (c *Config) Validate() error {
+	if c.Timeout <= 0 {
+		return fmt.Errorf("timeout must be greater than 0")
+	}
+	if c.Interval <= 0 {
+		return fmt.Errorf("interval must be greater than 0")
+	}
+	for _, t := range c.Targets {
+		if _, err := url.Parse(t.Addr); err != nil {
+			return fmt.Errorf("%s is not a valid url", t.Addr)
```

Comment:
Does it make sense checking for the scheme as well? http/https; or should all all protocols be allowed?

Reply:
There is no point in checking the scheme as http/Https is a layer 7 concern. We don't really care, the target just has to respond to a UDP packet with an icmp message. This will change a bit when moving to TCP, but the scheme is not a concern and should not be present. The only requirement is a valid IP or DNS name